apis/v1beta1/httproute_types.go

```diff
@@ -244,6 +244,56 @@ type HTTPRouteRule struct {
 	// +optional
 	// +kubebuilder:validation:MaxItems=16
 	BackendRefs []HTTPBackendRef `json:"backendRefs,omitempty"`
+
+	// Timeouts defines the timeouts that can be configured for an HTTP request.
+	//
+	// Support: Extended
+	//
+	// +optional
+	// <gateway:experimental>
+	Timeouts *HTTPRouteTimeouts `json:"timeouts,omitempty"`
+}
+
+// HTTPRouteTimeouts defines timeouts that can be configured for an HTTPRoute.
+// Timeout values are formatted like 1h/1m/1s/1ms as parsed by Golang time.ParseDuration
+// and MUST BE >= 1ms or 0 to disable (no timeout).
+type HTTPRouteTimeouts struct {
+	// Request specifies the duration for processing an HTTP client request after which the
+	// gateway will time out if unable to send a response.
+	//
+	// For example, setting the `rules.timeouts.request` field to the value `10s` in an
+	// `HTTPRoute` will cause a timeout if a client request is taking longer than 10 seconds
+	// to complete.
+	//
+	// This timeout is intended to cover as close to the whole request-response transaction
+	// as possible although an implementation MAY choose to start the timeout after the entire
+	// request stream has been received instead of immediately after the transaction is
+	// initiated by the client.
+	//
+	// When this field is unspecified, request timeout behavior is implementation-dependent.
+	//
+	// Support: Extended
+	//
+	// +optional
+	// +kubebuilder:validation:Format=duration
+	Request *metav1.Duration `json:"request,omitempty"`
+
+	// BackendRequest specifies a timeout for an individual request from the gateway
+	// to a backend service. This covers the time from when the request first starts being
+	// sent from the gateway to when the full response has been received from the backend.
+	//
+	// An entire client HTTP transaction with a gateway, covered by the Request timeout,
+	// may result in more than one call from the gateway to the destination backend service,
+	// for example, if automatic retries are supported.
+	//
+	// Because the Request timeout encompasses the BackendRequest timeout,
+	// the value of BackendRequest defaults to and must be <= the value of Request timeout.
```

Comment:
If we remove the "defaults to and" from here, I think we need to add another sentence to say the default value is implementation dependent.

That said, I thought that the result of the GEP review (where this comes from) was that defaulting to the value of the other field was sensible. Why do you think it's a problem? All this means is that if the `BackendRequest` field isn't set, then it default the the value that is set in the other fields (`Request`). The default value of `Request` in turn is implementation dependent (according to it's doc).

Reply:
Unfortunately we don't have a means to implement variable defaulting unless we want to create and maintain a defaulting webhook, and I'd strongly prefer to avoid that maintenance cost. 

We _could_ suggest that if a value is not set here, implementations should consider this the same value as the overall request timeout, but I can't think of any way to validate that, and I'm also not sure that's what we want. Let's say that an implementation has a default of 2 retries with a 10s per-try timeout. If a user configures an overall request timeout of 20s, I think they'd probably prefer the implementation default behavior here instead of that 1 value actually affecting the per-try timeout value as well. In general I'd like to suggest working towards the principle of least surprise, and I think that means avoiding a variable default like this.

Reply:
The way that I think about these timeouts composing means that I think BackendRequest can have a default of "no timeout".

If there is e.g. a 20s overall request timeout, I would expect the entire request to timeout after 20s even if no BackendRequest timeout is set.

Reply:
That's a great way of looking at this @adleong. I think we're all agreeing that there's no way for the value here to be greater than the overall request timeout value, even if it's not explicitly set. Maybe we should just focus on clearly stating that? That may have actually been what was originally intended here and I misunderstood it.

Reply:
The intent was as @adleong said, the request timeout applies if no backend timeout is set. I realize now that there is a subtle difference on how it's worded. To illustrate say we have a 10s Request timeout and 3 retries configured and we call a backend that is hanging. With the current wording (BackendRequest defaults to the value of Request) there will be 1 call to the backend which will timeout in 10s (no retries). If we change the wording to say BackendDefault is implementation dependent, then it's possible that an implementation has a shorter default backend timeout (say 2s) so all three retries will happen before the request fails in 8s. So if we remove the "defaults to Request" part, the default behavior when Request is set but BackendRequest is not, will be undefined. It's not really a big deal IMO, so I'm fine with it either way. Btw, Istio will work the BackendRequest default to Request way regardless.

Reply:
This became clearer to me in the other PR, but since implementations already have default behavior that is working with the API today, I don't think we can really add any kind of defaulting to these fields since it would result in backwards incompatible changes for implementations that already have defaults for per-try-timeout and/or retries.

So I think we should state that that this value can't be higher than request timeout since I don't think we can safely guarantee anything beyond that.